### ISSUE TITLE
Energy Conservation for STFT

### ISSUE THREAD

**Issue by DaanRateringRSM** [scipy/scipy#13489]
The project I am working on requires a short-time Fourier transformation. For this, I use scipy.signal.stft. From theory, we learn that the signal in time domain has the same energy as in time-frequency domain; Parseval's theorem (https://en.wikipedia.org/wiki/Parseval%27s_theorem). However, when I use scipy.signal.stft, I do not obtain the same amount of energy if I compare the computation for both values of energy. I am pretty sure, it is because there is some scaling involved when using scipy.signal.stft, so I spit through the source code. After some deep search, I have found a scaling factor of np.sqrt(1/(sum(win)**2)). However, using this scaling factor, does not work:

Example code:
```
fs = 10e3
N = 1e5
amp = 2 * np.sqrt(2)
noise_power = 0.01 * fs / 2
time = np.arange(N) / float(fs)
mod = 500*np.cos(2*np.pi*0.25*time)
carrier = amp * np.sin(2*np.pi*3e3*time + mod)
noise = np.random.normal(scale=np.sqrt(noise_power),
                         size=time.shape)
noise *= np.exp(-time/5)
x = carrier + noise

frame_size = 1000
window = scipy.signal.windows.hann(frame_size,sym=False)
f, t, Zxx = scipy.signal.stft(x, fs, window=window, nperseg=frame_size)
scale = np.sqrt(1/(sum(window)**2))
Zxx_scaled = Zxx/scale

Et=np.sum(x**2)
Ef=np.sum(abs(Zxx_scaled)**2)/N
print("Et = %.3f" % Et)
print( "Ef = %.3f" % Ef)
```

results in (depending on noise, but in the scale of):

```
Et = 1641328.607
Ef = 6158.539
```

Does anyone know how to properly scale the resulting STFT matrix values to obtain equal energy levels and prove energy conservation for the STFT?

(I am looking for a `sum(window)**2` constant transformation (and window) and, as the window is used in the STFT and I-STFT, I assume I have a 'tight frame' transformation if I use the hann window. However, proving energy conservation is a first step.....)

**Comment by larsoner** [comment]
I am not sure the answer to your question but recently I also tried to convert some [custom STFT code](https://github.com/mne-tools/mne-python/blob/24377ad3200b6099ed47576e9cf8b27578d571ef/mne/time_frequency/_stft.py#L64-L96) that [ensures the energy is conserved](https://github.com/mne-tools/mne-python/blob/24377ad3200b6099ed47576e9cf8b27578d571ef/mne/time_frequency/tests/test_stft.py#L23-L61) and was unable to do it with `scipy.signal.stft`'s current API. So it's possible there is some bug or some limitation in our current code that prevents properly conserving the energy.

It's also possible that there is some scale factor that could be computed and applied to get equivalence, in which case we need to document it properly.

**Comment by e-q** [comment]
It sounds like the scaling you mention is these lines:
https://github.com/scipy/scipy/blob/e555bf30a35076cfe06d19c1d3e02b62d0312fa8/scipy/signal/spectral.py#L1803-L1811
where `stft` is hard-coded to use `mode='spectrum'` in `_spectral_helper`. 

I wonder if the energy conservation problem only exists in the computation of one-sided STFTs, as we check the equivalence in the mean squared STFT magnitude to the power spectrum here, for the two-sided case:
https://github.com/scipy/scipy/blob/36ae01b16d6534c9e8c181959a33b8f6ff6a6f2b/scipy/signal/tests/test_spectral.py#L1158-L1176

**Comment by rgommers** [comment]
@larsoner as an FYI, I've cleared the milestone (not much is happening here, and it's not a release blocker).

**Comment by vollbier** [comment]
**tl;dr:** The internal function `_spectral_helper()` used by `stft()` can be parametrized to `scaling='density'` to provide the desired scaling. A solution might be to add that `scaling` parameter to `stft()`.

Dealing with spectral densities can be quite finicky, so for reference the following script compares `stf()` scaling to using the `rfft()` directly on a stationary signal:

```python
import numpy as np
from scipy import signal
from scipy.signal.spectral import _spectral_helper
import matplotlib.pyplot as plt


def my_stft(xx, fs=1.0, window='hann', nperseg=256, noverlap=None, nfft=None,
            detrend=False, return_onesided=True, boundary='zeros', padded=True,
            axis=-1, scaling='spectrum'):
    """This function differs to `signal.stft()` by exposing the `scaling` parameter. """
    freqs, time, Zxx = _spectral_helper(xx, xx, fs, window, nperseg, noverlap,
                                        nfft, detrend, return_onesided,
                                        scaling=scaling, axis=axis,
                                        mode='stft', boundary=boundary,
                                        padded=padded)

    return freqs, time, Zxx


# Generate a unit amplitude sine signal:
n, tau = 2**11, 10  # 16384 samples for a 10 s signal
m = 2**8  # window length
T = tau / n  # sampling interval

f0 = 50.4  # ensure to hit a frequency bin
t_x = np.arange(n) * T
x = np.sin(2 * np.pi * f0 * t_x)
print(f"The {f0} Hz sine signal is {tau} s long, consists of {n} samples, " +
      f"and has per definition the unit V.")
P_x = sum(x**2) / n  # signal power
print(f"The signal energy {P_x * tau:g} V²s is equal to power {P_x = :g} V² " +
      f"times duration {tau = :g} s.")

# Let's calculate a STFT with 'spectrum' and with `psd` scaling:
g = signal.windows.hann(m, sym=False)
kw = dict(fs=1 / T, window=g, nperseg=m, return_onesided=True)
f_stft, t_stft, Xs_stft = my_stft(x, scaling='spectrum', **kw)
Xd_stft = my_stft(x, scaling='density', **kw)[2]
a2_Xd_stft = Xd_stft.real ** 2 + Xd_stft.imag ** 2  # equals |Xd_stft|²


# For comparison, let's calculate a frequency line of the STFT directly with
# the rfft() and scale it to a magnitude spectrum or a PSD:
f_fft = np.fft.rfftfreq(m, T)
X_fft = np.fft.rfft(x[:m] * g)
X_mag = X_fft / (sum(abs(g)))  # unit area window
X_psd = X_fft * np.sqrt(T / sum(g ** 2))  # unit energy window
a2_X_psd = X_psd.real**2 + X_psd.imag**2

# Let's calculate the spectral power from the PSD:
df_fft = f_fft[1] - f_fft[0]
P_fft = np.sum(a2_X_psd) * df_fft
print(f"The one-sided spectral power {P_fft = :g} V² is half the signal power.")

# The spectral energy in signal can be calculated from the STFT as well:
df_stft, dt_stft = f_stft[1] - f_stft[0], t_stft[1] - t_stft[0]
Pt_stft = np.sum(a2_Xd_stft, axis=0) * df_stft  # integrate over f-axis
E_stft = np.sum(Pt_stft) * dt_stft  # integrate over t-axis
print(f"The energy of the one-sided STFT {E_stft = :g} V²s is half the signal energy.")


# Compare a frequency line of the STFT with the FFT:
fg2, (ax21, ax22) = plt.subplots(2, 1, sharex='all',  num=2, clear=True)
ax21.set(title="Magnitude Spectrum of $x(t)$ (unit is V)",
         ylabel="$|X(f)|$ in V", xlim=(0.9*f0, 1.1*f0))
ax22.set(title="Power spectral density", xlabel="Frequency $f$ in Hertz",
         ylabel=r"$\frac{d}{df}|X(f)|^2$ in V²/Hz")
k = len(t_stft) // 2  # pick a frequency line not to close to a border
ax21.plot(f_fft, abs(X_mag), 'C0-', alpha=.5, label="FFT")
ax21.plot(f_stft, abs(Xs_stft[:, k]), 'C1x--', alpha=.5, label="STFT (spectrum)")

ax22.plot(f_fft, a2_X_psd, 'C0-', alpha=.5, label="FFT")
ax22.plot(f_stft, a2_Xd_stft[:, k], 'C1x--', alpha=.5, label="STFT (density)")
for ax_ in (ax21, ax22):
    ax_.grid(True)
    ax_.legend()
plt.show()
```

__Please note:__

To ensure that the complete signal (energy) is captured by the STFT (more precisely: that it is a unitary mapping), the grid steps of the time-frequency grid need to be sufficiently small. E.g., `Δf Δt ≤ 1` is an upper bound. See Sec. 7.5 in [1] for more details.

Even though the `stft()` can be used to obtain information of a signal's energy distribution in the time-frequency plane by utilizing `density`-scaling, the naming is a bit unfortunate, since a STFT it cannot be interpreted as an energy *density*:

A density implies that its marginal, e.g., integrating over `f`, is the corresponding (temporal) power density `|x(t)|²`. Generally this is not the case for the STFT, due the window function not disappearing when integrating. The Wigner-Ville distribution has this marginalization property at the price of not guaranteeing non-negative values everywhere (which limits its usefulness considerably).

Furthermore, due to the Heisenberg uncertainty relation, a single point `(t_p, f_p)` in the time-frequency plane does not have a physical meaning — one has to look at areas of at least of size `1` to extract meaningful information about the energy distribution. See Sec 2 and Sec 3.3 in [1] for details.

__Literature__:
[1] Karlheinz Gröchenig, Foundations of Time-Frequency Analysis, Springer Science+Business Media, LLC, 2001.

**Comment by larsoner** [comment]
> The internal function _spectral_helper() used by stft() can be parametrized to scaling='density' to provide the desired scaling. A solution might be to add that scaling parameter to stft()... the naming is a bit unfortunate, since a STFT it cannot be interpreted as an energy density

It looks like `density` is already used publicly here:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.csd.html

If you think there is a better name for this mode, we could expose it and say that it's preferred, and keep `'density'` around just for backward compatibility.

@vollbier it sounds like you have a good handle on this issue, would you be up for making a PR? I think the steps would be:

1. Add `density` argument
2. Add unit tests that show
    1. That energy is preserved (i.e., Parseval's Theorem) when `density` is used, and
    2. That when using `scaling='density'` on both STFT and ISTFT, you get your original signal back (round-trip test)

Then a follow-up PR could be made after that to add an alias for `density` to all the functions that have a `density` option (e.g., welch, csd, stft, etc.). It would be great if one variant was adapted from the code at the top of the issue (assuming it executes quickly enough).

**Comment by vollbier** [comment]
@larsoner Concerning the naming `{ 'density', 'spectrum' }` in [csd() ](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.csd.html) and friends:   Densities in pure spectral  (or pure time) domains are unproblematic—so no problem there. Things become only hairy when looking at _time-frequency_ distributions.

For the case of the STFT, a frequency line (i.e., fixed `t`) may be interpreted as a power spectral density (psd), but a  time-line (i.e., fixed `f`) is not a density function.  Hence, I would suggest to replace `density` with  `psd` only there.

If nobody else volunteers, I'd be happy to give a shot at writing a PR.

**Comment by larsoner** [comment]
@vollbier this issue has been open for a while and nobody has worked on it, so feel free to open a pull request whenever you make some progress!
